docs/proposals/default-service-plans.md

```diff
@@ -0,0 +1,544 @@
+* **{{ service-name }}-server-only**: Represents a database management system or server only. Example: mysql-server-only
+* **{{ service-name }}-database-and-server**: Represents both a database server and a logical database. Example: mysql-database-and-server
+* **{{ service-name }}**: Represents a service where there are not parent-child relationships between service instances. _Parent-child relationships are a new concept in OSB that is still being discussed._ Example: redis, rabbitmq
+
+
+#### Consumer
+
+End users should be able to install a chart on a cluster with any broker that supports this proposal, and not be required to specify any broker-specific values:
+
+
+``` console
+$ helm install azure/wordpress
+# Provision a new service of type mysql and inject it into Wordpress
+```
+
+
+If the cluster operator has created custom classes, that can be specified when installing the chart:
+
+
+``` console
+$ helm install azure/wordpress \
+    --set className=custom-mysql --set planName=ha-mysql
+```
+
+
+
+## Implementation
+
+A key goal of the implementation is to avoid going off and rewriting or requiring a long term branch or fork, etc. I'd like to see this developed using a feature flag, and making all new fields optional.
+
+
+### Service Class
+
+
+| Field | Description |
+|-------|-------------|
+| Service Type | Examples: mysql, redis|
+| <p>Default Provision Parameters</p><p>Default Bind Parameters</p> | <p>Default parameters to supply to the broker during provisioning and binding.</p> <p>OSB API V3 - Make the schema required, and defaults optional.|
+| Default Bind Response Transform | <p>Default json key mapping to transform a non-standard broker response into something that can be relied upon.</p> <p>For example, if a broker returns "database.name" (a nested value) in the response, it can be changed to "database" to un-nest it and apply a standard name.</p>|
+| Controller Reference |<p>Set by Service Catalog when populating the catalog from a broker list to indicate that it is managed by the controller.</p> <p>When not present, the class is ignored during a relist. This gives operators a way to define custom classes.</p>|
+
+Notes
+
+* Relist needs to take into account if the class is managed by the controller or was created by the cluster operator.
+* Defaults can be defined at the Cluster or Namespace level.
+* For simplification, only broker bind response bodies can be transformed, not provision, since that's where we require that functionality today.
+
+
+### Service Plan
+
+| Field | Description |
+|-------|-------------|
+| Service Type | Examples: mysql, redis |
+| annotation: is-default-plan=true | <p>Specifies whether or not a Service Plan is the default for its Service Type.</p> <p>When annotated as the default and a Service Type is set, this plan is used to dynamically provision an instance for that type.</p> |
```

Comment:
This feels like it should be moved to the service class definition. Otherwise it is too easy to have two plans with this annotation set to true

Reply:
I was hoping that we could have the controller enforce that only 1 plan has the default set to true for a particular service type. Do you think that is feasible?

Reply:
For example:

Let's say that there are two brokers installed on the cluster that both provide a MySQL database and the operator hasn't picked which one should be the default. When the developer installs a helm chart that needs a MySQL database, it will fail. The status message on the ServiceInstance will indicate that there is more than one default plan for MySQL, and that either the cluster operator should pick a default and then touch the service instance (to retrigger provisioning), or you edit the service instance to select specific class+plan.

Here's more information on [how the default plan is resolved](https://github.com/carolynvs/service-catalog/blob/default-service-plan-proposal/docs/proposals/default-service-plans.md#default-plan-resolution).